subprojects/execution/src/main/java/org/gradle/internal/execution/UnitOfWork.java

```diff
@@ -42,13 +45,35 @@
 
     InputChangeTrackingStrategy getInputChangeTrackingStrategy();
 
+    void visitImplementations(ImplementationVisitor visitor);
+
+    interface ImplementationVisitor {
+        void visitImplementation(Class<?> implementation);
+        void visitImplementation(ImplementationSnapshot implementation);
+        void visitAdditionalImplementation(Class<?> implementation);
+        void visitAdditionalImplementation(ImplementationSnapshot implementation);
+    }
+
+    void visitInputProperties(InputPropertyVisitor visitor);
+
+    interface InputPropertyVisitor {
+        void visitInputProperty(String propertyName, Object value);
+    }
+
     void visitInputFileProperties(InputFilePropertyVisitor visitor);
 
+    interface InputFilePropertyVisitor {
+        void visitInputFileProperty(String propertyName, @Nullable Object value, boolean incremental, Supplier<CurrentFileCollectionFingerprint> fingerprinter);
+    }
+
     void visitOutputProperties(OutputPropertyVisitor visitor);
 
+    interface OutputPropertyVisitor {
+        void visitOutputProperty(String propertyName, TreeType type, FileCollection roots);
+    }
+
     void visitLocalState(LocalStateVisitor visitor);
 
-    @FunctionalInterface
```

Comment:
Why did you remove the `@FunctionalInterface` annotations?

Reply:
The fact that these interfaces can currently be implemented as functions is basically just a coincidence, and not something I intend to keep like that necessarily.

The visitors are not supposed to be implemented by the consumers of this API, so they shouldn't care much.

The only use I see of this annotation is that the compiler would tell me if I accidentally(?) added another method. Which doesn't sound very useful in this situation. So I opted to remove the noise.